File: utilities/spellcasting.py
# spellcasting.py
# Spell casting utility, separated from main.py
# Progressing through decentralisation...

import random
from typing import Dict, List, Any, Optional
from utilities.settings import Colors
import utilities.dice


class SpellCastingSystem:
    """System for handling spell casting"""
    
    def __init__(self, game_instance):
        self.game = game_instance
        self.player = game_instance.player
        self.lang = game_instance.lang
        self.spells_data = game_instance.spells_data
        self.effects_data = game_instance.effects_data
        self.items_data = game_instance.items_data
        self.dice_util = utilities.dice.Dice()
    
    def get_available_spells(self, weapon_name: Optional[str]) -> List[tuple]:
        """Get available spells for a weapon"""
        if not weapon_name:
            return []
        
        available = []
        for sname, sdata in self.spells_data.items():
            allowed = sdata.get('allowed_weapons', [])
            if weapon_name in allowed:
                available.append((sname, sdata))
        return available
    
    def can_cast_spells(self, weapon_name: Optional[str]) -> bool:
        """Check if a weapon can cast spells"""
        if not weapon_name:
            return False
        weapon_data = self.items_data.get(weapon_name, {})
        return bool(weapon_data.get('magic_weapon'))
    
    def cast_spell(self, enemy, spell_name: str, spell_data: Dict[str, Any]) -> Dict[str, Any]:
        """Cast a spell on an enemy"""
        if not self.player:
            return {'success': False, 'error': 'No player'}
        
        cost = spell_data.get('mp_cost', 0)
        
        if self.player.mp < cost:
            print(self.lang.get("not_enough_mp", "Not enough MP!"))
            return {'success': False, 'error': 'Not enough MP'}
        
        # Pay cost
        self.player.mp -= cost
        
        spell_type = spell_data.get('type')
        result = {'success': True, 'type': spell_type, 'spell_name': spell_name}
        
        if spell_type == 'damage':
            self._cast_damage_spell(enemy, spell_name, spell_data, result)
        elif spell_type == 'heal':
            self._cast_heal_spell(spell_name, spell_data, result)
        elif spell_type == 'buff':
            self._cast_buff_spell(spell_name, spell_data, result)
        elif spell_type == 'debuff':
            self._cast_debuff_spell(enemy, spell_name, spell_data, result)
        else:
            print(f"Unknown spell type: {spell_type}")
            # Refund MP for unknown spell types
            self.player.mp += cost
            result['success'] = False
            result['error'] = 'Unknown spell type'
        
        # Check for cast cutscene
        cast_cutscene = spell_data.get('cast_cutscene')
        if cast_cutscene and hasattr(self.game, 'cutscenes_data') and cast_cutscene in self.game.cutscenes_data:
            if hasattr(self.game, 'play_cutscene'):
                self.game.play_cutscene(cast_cutscene)
        
        return result
    
    def _cast_damage_spell(self, enemy, spell_name: str, spell_data: Dict[str, Any], result: Dict[str, Any]):
        """Cast a damage spell"""
        power = spell_data.get('power', 0)
        base_damage = power + (self.player.get_effective_attack() // 2)
        
        roll = self.dice_util.roll_1d(20)
        
        if roll == 1:
            meme_key = f"roll_1_meme_{random.randint(1, 3)}"
            print(self.lang.get(meme_key))
        elif roll == 20:
            meme_key = f"roll_20_meme_{random.randint(1, 3)}"
            print(self.lang.get(meme_key))
        else:
            print(self.lang.get("roll_msg", roll=roll))
        
        damage = int(base_damage * roll / 10)
        actual = enemy.take_damage(damage)
        
        print(f"You cast {spell_name} for {actual} damage!")
        result['damage'] = actual
        
        # Apply effects if any
        effects = spell_data.get('effects', [])
        for effect_name in effects:
            effect_data = self.effects_data.get(effect_name, {})
            effect_type = effect_data.get('type', '')
            
            if effect_type == 'damage_over_time':
                print(f"{Colors.RED}{enemy.name} is afflicted with {effect_name}!{Colors.END}")
            elif effect_type == 'stun':
                if random.random() < effect_data.get('chance', 0.5):
                    print(f"{Colors.YELLOW}{enemy.name} is stunned!{Colors.END}")
            elif effect_type == 'mixed_effect':
                if random.random() < effect_data.get('chance', 0.5):
                    print(f"{Colors.CYAN}{enemy.name} is frozen!{Colors.END}")
    
    def _cast_heal_spell(self, spell_name: str, spell_data: Dict[str, Any], result: Dict[str, Any]):
        """Cast a heal spell"""
        heal_amount = spell_data.get('power', 0)
        old_hp = self.player.hp
        self.player.heal(heal_amount)
        healed = self.player.hp - old_hp
        
        print(f"You cast {spell_name} and healed {healed} HP!")
        result['healed'] = healed
        
        # Apply healing effects if any
        effects = spell_data.get('effects', [])
        for effect_name in effects:
            effect_data = self.effects_data.get(effect_name, {})
            if effect_data.get('type') == 'healing_over_time':
                print(f"{Colors.GREEN}You are affected by regeneration!{Colors.END}")
    
    def _cast_buff_spell(self, spell_name: str, spell_data: Dict[str, Any], result: Dict[str, Any]):
        """Cast a buff spell"""
        power = spell_data.get('power', 0)
        effects = spell_data.get('effects', [])
        
        for effect_name in effects:
            effect_data = self.effects_data.get(effect_name, {})
            effect_type = effect_data.get('type', '')
            
            # Collect numeric modifiers from effect_data
            modifiers = {}
            for k, v in effect_data.items():
                if isinstance(v, (int, float)) and (
                    k.endswith('_bonus') or
                    k in ('hp_bonus', 'mp_bonus', 'absorb_amount', 'critical_bonus')
                ):
                    modifiers[k] = int(v)
            
            duration = int(effect_data.get('duration', max(3, power or 3)))
            
            # Apply as temporary buff
            if modifiers:
                self.player.apply_buff(effect_name, duration, modifiers)
                mod_str = ', '.join(f"{v} {k}" for k, v in modifiers.items())
                print(f"{Colors.GREEN}Applied buff: {effect_name} (+{mod_str}) for {duration} turns{Colors.END}")
            else:
                # Non-numeric effects still applied as a marker buff
                self.player.apply_buff(effect_name, duration, {})
                if effect_type == 'damage_absorb':
                    print(f"{Colors.BLUE}You create a magical shield!{Colors.END}")
                elif effect_type == 'reconnaissance':
                    print(f"{Colors.CYAN}You can see enemy weaknesses!{Colors.END}")
        
        result['buffs_applied'] = len(effects)
    
    def _cast_debuff_spell(self, enemy, spell_name: str, spell_data: Dict[str, Any], result: Dict[str, Any]):
        """Cast a debuff spell"""
        power = spell_data.get('power', 0)
        effects = spell_data.get('effects', [])
        
        for effect_name in effects:
            effect_data = self.effects_data.get(effect_name, {})
            effect_type = effect_data.get('type', '')
            
            if effect_type == 'action_block':
                if random.random() < effect_data.get('chance', 0.5):
                    print(f"{Colors.YELLOW}{enemy.name} is stunned and cannot act!{Colors.END}")
            elif effect_type == 'accuracy_reduction':
                print(f"{Colors.RED}{enemy.name}'s accuracy is reduced!{Colors.END}")
            elif effect_type == 'speed_reduction':
                print(f"{Colors.YELLOW}{enemy.name} is slowed!{Colors.END}")
            elif effect_type == 'stat_reduction':
                print(f"{Colors.RED}{enemy.name}'s stats are cursed!{Colors.END}")
        
        result['debuffs_applied'] = len(effects)
    
    def select_spell(self, weapon_name: Optional[str]) -> Optional[tuple]:
        """Display spell selection menu and return selected spell"""
        from main import clear_screen, ask
        
        available = self.get_available_spells(weapon_name)
        
        if not available:
            print(self.lang.get("no_spells_available", "No spells available for this weapon."))
            return None
        
        page = 0
        per_page = 10
        
        while True:
            clear_screen()
            total_pages = max(1, (len(available) + per_page - 1) // per_page)
            start_idx = page * per_page
            end_idx = start_idx + per_page
            current_spells = available[start_idx:end_idx]
            
            print(f"\n{Colors.BOLD}=== SPELLS (Page {page + 1}/{total_pages}) ==={Colors.END}")
            print(f"MP: {Colors.BLUE}{self.player.mp}/{self.player.max_mp}{Colors.END}\n")
            
            for i, (sname, sdata) in enumerate(current_spells, 1):
                cost = sdata.get('mp_cost', 0)
                mp_color = Colors.BLUE if self.player.mp >= cost else Colors.RED
                print(f"{i}. {Colors.CYAN}{sname}{Colors.END} - Cost: {mp_color}{cost} MP{Colors.END}")
                print(f"   {sdata.get('description', '')}")
            
            print("\nOptions:")
            if total_pages > 1:
                if page > 0:
                    print(f"P. {self.lang.get('ui_previous_page', 'Previous Page')}")
                if page < total_pages - 1:
                    print(f"N. {self.lang.get('ui_next_page', 'Next Page')}")
            
            print(f"1-{len(current_spells)}. Cast Spell")
            print(f"B. {self.lang.get('back', 'Back')}")
            
            choice = ask("\nChoose an option: ").strip().upper()
            
            if choice == 'B' or not choice:
                return None
            elif choice == 'N' and page < total_pages - 1:
                page += 1
            elif choice == 'P' and page > 0:
                page -= 1
            elif choice.isdigit():
                idx = int(choice) - 1
                if 0 <= idx < len(current_spells):
                    sname, sdata = current_spells[idx]
                    return (sname, sdata)
                else:
                    print(self.lang.get('invalid_selection', "Invalid selection"))
                    import time
                    time.sleep(1)
            else:
                print(self.lang.get("invalid_choice", "Invalid choice"))
                import time
                time.sleep(1)

File: utilities_js/spellcasting.js
// spellcasting.js
// Spell casting utility for browser, separated from main game logic
// Progressing through decentralisation...
// Andy64lol

import { Dice } from './dice.js';
import { Colors } from './settings.js';

/**
 * SpellCastingSystem class for handling spell casting
 * Ported from utilities/spellcasting.py
 */
export class SpellCastingSystem {
  /**
   * @param {Object} gameInstance - The game instance with player, lang, data references
   */
  constructor(gameInstance) {
    this.game = gameInstance;
    this.player = gameInstance.player;
    this.lang = gameInstance.lang;
    this.spellsData = gameInstance.spellsData;
    this.effectsData = gameInstance.effectsData;
    this.itemsData = gameInstance.itemsData;
    this.diceUtil = new Dice();
  }

  /**
   * Get available spells for a weapon
   * @param {string} weaponName - The weapon name to check
   * @returns {Array} Array of [spellName, spellData] tuples
   */
  getAvailableSpells(weaponName) {
    if (!weaponName) {
      return [];
    }

    const available = [];
    for (const [sname, sdata] of Object.entries(this.spellsData)) {
      const allowed = sdata.allowedWeapons || sdata.allowed_weapons || [];
      if (allowed.includes(weaponName)) {
        available.push([sname, sdata]);
      }
    }
    return available;
  }

  /**
   * Check if a weapon can cast spells
   * @param {string} weaponName - The weapon name to check
   * @returns {boolean} True if weapon can cast spells
   */
  canCastSpells(weaponName) {
    if (!weaponName) return false;
    const weaponData = this.itemsData[weaponName] || {};
    return Boolean(weaponData.magicWeapon || weaponData.magic_weapon);
  }

  /**
   * Cast a spell on an enemy
   * @param {Object} enemy - The enemy to cast spell on
   * @param {string} spellName - The name of the spell to cast
   * @param {Object} spellData - The spell data
   * @returns {Object} Result of the spell cast
   */
  castSpell(enemy, spellName, spellData) {
    if (!this.player) {
      return { success: false, error: 'No player' };
    }

    const cost = spellData.mpCost || spellData.mp_cost || 0;
    
    if (this.player.mp < cost) {
      this.game.print(this.lang.get("not_enough_mp", "Not enough MP!"));
      return { success: false, error: 'Not enough MP' };
    }

    // Pay cost
    this.player.mp -= cost;

    const spellType = spellData.type;
    const result = { success: true, type: spellType, spellName };

    switch (spellType) {
      case 'damage':
        this._castDamageSpell(enemy, spellName, spellData, result);
        break;
      case 'heal':
        this._castHealSpell(spellName, spellData, result);
        break;
      case 'buff':
        this._castBuffSpell(spellName, spellData, result);
        break;
      case 'debuff':
        this._castDebuffSpell(enemy, spellName, spellData, result);
        break;
      default:
        this.game.print(`Unknown spell type: ${spellType}`);
        // Refund MP for unknown spell types
        this.player.mp += cost;
        result.success = false;
        result.error = 'Unknown spell type';
    }

    // Check for cast cutscene
    const castCutscene = spellData.castCutscene || spellData.cast_cutscene;
    if (castCutscene && this.game.cutscenesData && this.game.cutscenesData[castCutscene]) {
      if (typeof this.game.playCutscene === 'function') {
        this.game.playCutscene(castCutscene);
      }
    }

    return result;
  }

  /**
   * Cast a damage spell
   * @private
   */
  _castDamageSpell(enemy, spellName, spellData, result) {
    const power = spellData.power || 0;
    const baseDamage = power + Math.floor(this.player.getEffectiveAttack() / 2);
    
    const roll = this.diceUtil.roll_1d(20);
    
    if (roll === 1) {
      const memeKey = `roll_1_meme_${Math.floor(Math.random() * 3) + 1}`;
      this.game.print(this.lang.get(memeKey));
    } else if (roll === 20) {
      const memeKey = `roll_20_meme_${Math.floor(Math.random() * 3) + 1}`;
      this.game.print(this.lang.get(memeKey));
    } else {
      this.game.print(this.lang.get("roll_msg", { roll: roll }));
    }

    const damage = Math.floor(baseDamage * roll / 10);
    const actual = enemy.takeDamage(damage);
    
    this.game.print(`You cast ${spellName} for ${actual} damage!`);
    result.damage = actual;

    // Apply effects if any
    const effects = spellData.effects || [];
    for (const effectName of effects) {
      const effectData = this.effectsData[effectName] || {};
      const effectType = effectData.type || '';

      if (effectType === 'damage_over_time') {
        this.game.print(`${enemy.name} is afflicted with ${effectName}!`);
      } else if (effectType === 'stun') {
        if (Math.random() < (effectData.chance || 0.5)) {
          this.game.print(`${enemy.name} is stunned!`);
        }
      } else if (effectType === 'mixed_effect') {
        if (Math.random() < (effectData.chance || 0.5)) {
          this.game.print(`${enemy.name} is frozen!`);
        }
      }
    }
  }

  /**
   * Cast a heal spell
   * @private
   */
  _castHealSpell(spellName, spellData, result) {
    const healAmount = spellData.power || 0;
    const oldHp = this.player.hp;
    this.player.heal(healAmount);
    const healed = this.player.hp - oldHp;
    
    this.game.print(`You cast ${spellName} and healed ${healed} HP!`);
    result.healed = healed;

    // Apply healing effects if any
    const effects = spellData.effects || [];
    for (const effectName of effects) {
      const effectData = this.effectsData[effectName] || {};
      if (effectData.type === 'healing_over_time') {
        this.game.print(`You are affected by regeneration!`);
      }
    }
  }

  /**
   * Cast a buff spell
   * @private
   */
  _castBuffSpell(spellName, spellData, result) {
    const power = spellData.power || 0;
    const effects = spellData.effects || [];

    for (const effectName of effects) {
      const effectData = this.effectsData[effectName] || {};
      const effectType = effectData.type || '';

      // Collect numeric modifiers from effect_data
      const modifiers = {};
      for (const [k, v] of Object.entries(effectData)) {
        if (typeof v === 'number' && (
          k.endsWith('_bonus') || k.endsWith('Bonus') ||
          ['hp_bonus', 'mp_bonus', 'absorb_amount', 'critical_bonus'].includes(k)
        )) {
          modifiers[k] = Math.floor(v);
        }
      }

      const duration = Math.floor(effectData.duration || Math.max(3, power || 3));
      
      // Apply as temporary buff
      if (Object.keys(modifiers).length > 0) {
        this.player.applyBuff(effectName, duration, modifiers);
        const modStr = Object.entries(modifiers).map(([k, v]) => `${v} ${k}`).join(', ');
        this.game.print(`Applied buff: ${effectName} (+${modStr}) for ${duration} turns`);
      } else {
        // Non-numeric effects still applied as a marker buff
        this.player.applyBuff(effectName, duration, {});
        if (effectType === 'damage_absorb') {
          this.game.print(`You create a magical shield!`);
        } else if (effectType === 'reconnaissance') {
          this.game.print(`You can see enemy weaknesses!`);
        }
      }
    }
    
    result.buffsApplied = effects.length;
  }

  /**
   * Cast a debuff spell
   * @private
   */
  _castDebuffSpell(enemy, spellName, spellData, result) {
    const power = spellData.power || 0;
    const effects = spellData.effects || [];

    for (const effectName of effects) {
      const effectData = this.effectsData[effectName] || {};
      const effectType = effectData.type || '';

      if (effectType === 'action_block') {
        if (Math.random() < (effectData.chance || 0.5)) {
          this.game.print(`${enemy.name} is stunned and cannot act!`);
        }
      } else if (effectType === 'accuracy_reduction') {
        this.game.print(`${enemy.name}'s accuracy is reduced!`);
      } else if (effectType === 'speed_reduction') {
        this.game.print(`${enemy.name} is slowed!`);
      } else if (effectType === 'stat_reduction') {
        this.game.print(`${enemy.name}'s stats are cursed!`);
      }
    }
    
    result.debuffsApplied = effects.length;
  }

  /**
   * Display spell selection menu (for CLI/browser use)
   * @param {string} weaponName - The weapon to get spells for
   * @returns {Promise<Object|null>} Selected spell or null if cancelled
   */
  async selectSpell(weaponName) {
    const available = this.getAvailableSpells(weaponName);
    
    if (available.length === 0) {
      this.game.print(this.lang.get("no_spells_available", "No spells available for this weapon."));
      return null;
    }

    let page = 0;
    const perPage = 10;

    while (true) {
      const totalPages = Math.ceil(available.length / perPage);
      const start = page * perPage;
      const currentSpells = available.slice(start, start + perPage);

      this.game.print(`\n=== SPELLS (Page ${page + 1}/${totalPages}) ===`);
      this.game.print(`MP: ${this.player.mp}/${this.player.maxMp}\n`);

      for (let i = 0; i < currentSpells.length; i++) {
        const [sname, sdata] = currentSpells[i];
        const cost = sdata.mpCost || sdata.mp_cost || 0;
        this.game.print(`${i + 1}. ${sname} - Cost: ${cost} MP`);
        this.game.print(`   ${sdata.description || ''}`);
      }

      this.game.print("\nOptions:");
      if (totalPages > 1) {
        if (page > 0) this.game.print("P. Previous Page");
        if (page < totalPages - 1) this.game.print("N. Next Page");
      }
      this.game.print(`1-${currentSpells.length}. Cast Spell`);
      this.game.print("B. Back");

      // In browser, this would use a prompt or UI element
      const choice = await this.game.ask("Choose an option: ");
      if (!choice) return null;
      
      const upperChoice = choice.toUpperCase();

      if (upperChoice === 'B') {
        return null;
      } else if (upperChoice === 'N' && page < totalPages - 1) {
        page++;
      } else if (upperChoice === 'P' && page > 0) {
        page--;
      } else if (!isNaN(parseInt(choice))) {
        const idx = parseInt(choice) - 1;
        if (idx >= 0 && idx < currentSpells.length) {
          const [sname, sdata] = currentSpells[idx];
          return { name: sname, data: sdata };
        } else {
          this.game.print(this.lang.get('invalid_selection', "Invalid selection"));
        }
      } else {
        this.game.print(this.lang.get("invalid_choice", "Invalid choice"));
      }
    }
  }
}
